tests/functional/test_model_backcompat.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import os
+
+from nose.tools import assert_equal
+from botocore.session import Session
+from botocore.stub import Stubber
+
+
+FIXED_MODELS_DIR = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), 'models',
+)
+
+
+def test_old_model_continues_to_work():
+    # This test ensures that botocore can load the service models as they exist
+    # today.  There's a directory in tests/functional/models that is a
+    # snapshot of a service model.  This test ensures that we can continue
+    # to stub an API call using this model.  That way if the models ever
+    # change we have a mechanism to ensure that the existing models continue
+    # to work with botocore.  The test should not change, and the files in
+    # tests/functional/models should not change!
+    session = Session()
+    loader = session.get_component('data_loader')
+    # We're adding our path to the existing search paths so we don't have to
+    # copy additional data files such as _retry.json to our FIXED_MODELS_DIR.
+    # We only care about the service model and endpoints file not changing.
+    # This also prevents us from having to make any changes to this models dir
+    # if we end up adding a new data file that's needed to create clients.
+    # We're adding our FIXED_MODELS_DIR as the first element in the list to
+    # ensure we load the endpoints.json file from FIXED_MODELS_DIR.  For the
+    # service model we have an extra safety net where we can choose a custom
+    # client name.
+    loader.search_paths.insert(0, FIXED_MODELS_DIR)
+
+    # The model dir we copied was renamed to 'custom-lambda'
+    # to ensure we're loading our version of the model and not
+    # the built in one.
+    client = session.create_client(
+        'custom-lambda', region_name='us-west-2',
+        aws_access_key_id='foo', aws_secret_access_key='bar',
+    )
+    with Stubber(client) as stubber:
```

Comment:
While I agree I don't think we need to be making an actual API request, what is your thoughts on using the [ClientHTTPStubber](https://github.com/boto/botocore/blob/334a179078e71760cfc8487b129ae894f886d6a0/tests/__init__.py#L380) instead? The nice part about this is we will cover a bit more in our stack and still will not be making an API call over the wire.

Reply:
In general I'm not opposed but I'd prefer to stick with the stubber for now because it's a publicly documented API that will remain backwards compatible.  Ideally we're not changing this test so I tried to stick with APIs that should never require modifications.

If we move the `ClientHTTPStubber` over to the stub module, then perhaps, but it sounds like there's some more vetting we'd like to do with the `ClientHTTPStubber` before making it public.

Reply:
Yeah I agree we would want to do more vetting before making it public. While I do not think we should be touching this test very much, I can see us updating it if we add support for a new model and if we did update the `ClientHTTPStubber`, I'm not sure if it would be a significant update to the interface. I think the main thing I'm concerned about is that the stubber is registered near the beginning of the `before-call` event (via `before-call.*.*`), which is a popular event we use. So if we registered a handler after the stubber handler and it caused issues for the legacy models, we may be blind to it because the stubber causes all subsequently registered handlers to not get ran.